engine/rpcserver.go

```diff
@@ -1213,3 +1220,283 @@ func (s *RPCServer) GetAuditEvent(ctx context.Context, r *gctrpc.GetAuditEventRe
+	}
+
+	gctVM := gctscript.New()
+	if gctVM == nil {
+		return &gctrpc.GCTScriptGenericResponse{}, nil
+	}
+
+	script := filepath.Join(r.Script.Path, r.Script.Name)
+	err := gctVM.Load(script)
+	if err != nil {
+		return &gctrpc.GCTScriptGenericResponse{
+			Status: "error",
+			Data:   err.Error(),
+		}, nil
+	}
+
+	go gctVM.CompileAndRun()
+
+	return &gctrpc.GCTScriptGenericResponse{
+		Status: "ok",
+		Data:   gctVM.ShortName() + " (" + gctVM.ID.String() + ") executed",
+	}, nil
+}
+
+// GCTScriptStop terminate a running script
+func (s *RPCServer) GCTScriptStop(ctx context.Context, r *gctrpc.GCTScriptStopRequest) (*gctrpc.GCTScriptGenericResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	UUID, err := uuid.FromString(r.Script.UUID)
+	if err != nil {
+		return &gctrpc.GCTScriptGenericResponse{Status: "error", Data: err.Error()}, nil
+	}
+
+	if v, f := gctscript.AllVMs[UUID]; f {
+		err = v.Shutdown()
+		status := " terminated"
+		if err != nil {
+			status = err.Error()
+		}
+		return &gctrpc.GCTScriptGenericResponse{Status: "ok", Data: v.ID.String() + status}, nil
+	}
+	return &gctrpc.GCTScriptGenericResponse{Status: "error", Data: "no running script found"}, nil
+}
+
+// GCTScriptUpload upload a new script to ScriptPath
+func (s *RPCServer) GCTScriptUpload(ctx context.Context, r *gctrpc.GCTScriptUploadRequest) (*gctrpc.GCTScriptGenericResponse, error) {
+	if !gctscript.GCTScriptConfig.Enabled {
+		return &gctrpc.GCTScriptGenericResponse{Status: gctscript.ErrScriptingDisabled.Error()}, nil
+	}
+
+	fPath := filepath.Join(gctscript.ScriptPath, r.ScriptName)
+	_, err := os.Stat(fPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+	} else {
+		if !r.Overwrite {
+			return nil, fmt.Errorf("%s script found and overwrite set to false", r.ScriptName)
+		} else if r.Overwrite {
+			f := filepath.Join(gctscript.ScriptPath, "version_history")
+			err = os.MkdirAll(f, 0770)
+			if err != nil {
+				return nil, err
+			}
+			renamedFile := filepath.Join(f, time.Now().Format(time.RFC3339)+" "+r.ScriptName)
+			err = file.Move(fPath, renamedFile)
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	newFile, err := os.Create(fPath)
+	if err != nil {
+		return nil, err
+	}
+	defer newFile.Close()
+
+	_, err = newFile.Write(r.Data)
+	if err != nil {
+		return nil, err
+	}
+
+	err = gctscript.Validate(fPath)
+	if err != nil {
+		errRemove := os.Remove(fPath)
+		if errRemove != nil {
+			return nil, err
+		}
+		return &gctrpc.GCTScriptGenericResponse{Status: "validation failed", Data: err.Error()}, nil
+	}
+
+	if r.Archived {
+		z, err := zip.OpenReader(fPath)
+		if err != nil {
+			return nil, err
+		}
+
+		defer z.Close()
+
+		for x := range z.Reader.File {
+			var zFile io.ReadCloser
+			zFile, err = z.Reader.File[x].Open()
+			if err != nil {
+				return nil, err
+			}
+			defer zFile.Close()
```

Comment:
Since we're in a for loop, this will only close upon func exit. It would be better to manually close through each iteration of the for loop rather than keep these file handles all open until then